docs/user-guide/materialization.md

```diff
@@ -0,0 +1,236 @@
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+```
+
+To materialize the nix files we need to take care to pin down the
+inputs.  For `cabalProject` and `hackage-package` this means
+we must specify the `index-state` of hackage we want to use:
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+  index-state = "2019-12-03T00:00:00Z";
+}).components.exes.hlint
+```
+
+Now if we build again we get a hint telling use how to
+calculate a suitable sha256 hash to turn the derivation
+containing the nix files into a fixed output derivation:
+
+```
+$ nix-build hlint.nix 
+trace: Using index-state: 2019-12-03T00:00:00Z for hlint
+trace: Get `plan-sha256` with `nix-hash --base32 --type sha256 /nix/store/qk1fvza1alkvs51vzmpjp2xsg8xklyxk-hlint-plan-to-nix-pkgs/`
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+$ nix-hash --base32 --type sha256 /nix/store/qk1fvza1alkvs51vzmpjp2xsg8xklyxk-hlint-plan-to-nix-pkgs/
+1a4rhv3h2daz6dzwzfl3w7l1v556n7aqfiagw6m0rvqq230iabss
+```
+
+We can add the hash as `plan-sha256` or (`stack-sha256` for
+`stackProject`)
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+  index-state = "2019-12-03T00:00:00Z";
+  plan-sha256 = "1a4rhv3h2daz6dzwzfl3w7l1v556n7aqfiagw6m0rvqq230iabss";
+}).components.exes.hlint
+```
+
+Just adding the hash might help reuse of the cached nix, but nix will
+still calculate all the dependencies (which can add seconds to
+`nix-build` and `nix-shell` commands when no ohter work is needed)
+and users who do not yet have the dependencies in their store will have
+to wait while they are built or downloaded.
+
+Running nix build again gives us a hint on what we can do next:
+
+```
+$ nix-build hlint.nix 
+trace: Using index-state: 2019-12-03T00:00:00Z for hlint
+trace: To materialize copy /nix/store/0xalcphb7ifvy5fc9dpwj40fij6nn5av-hlint-plan-to-nix-pkgs
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+```
+
+To capture the nix we can do something like:
+
+```nix
+((import ./nixpkgs (import ./.)).haskell-nix.hackage-package {
+  name = "hlint";
+  version = "2.2.4";
+  index-state = "2019-12-03T00:00:00Z";
+  plan-sha256 = "1a4rhv3h2daz6dzwzfl3w7l1v556n7aqfiagw6m0rvqq230iabss";
+  materialized = ./hlint.materialized;
+}).components.exes.hlint
+```
+
+Now we can copy the nix files needed and build with:
+
+```
+$ cp -r /nix/store/0xalcphb7ifvy5fc9dpwj40fij6nn5av-hlint-plan-to-nix-pkgs hlint.materialized
+$ nix-build hlint.nix 
+/nix/store/7wwqq5v42gm6iiz2d3ngbnkyw7i4py11-hlint-2.2.4-exe-hlint
+```
+
+We may want to run `cmod -R +w hlint.materialized` as the files copied from the
+store will be read only.
+
+## How can we check `sha256` and `materialized` are up to date?
+
+Let's pretend we had to go back to `hlint` version `2.2.3`.
+We can change `version` and temporarily add
+`checkMaterialization = true;`:
```

Comment:
We should note why one would not always want to set `checkMaterialization = true`. Performance?

Reply:
Done

Reply:
(At the bottom of the section)